e2e/testcases/ssh_known_hosts_test.go

```diff
@@ -0,0 +1,119 @@
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/reconcilermanager/controllers"
+)
+
+func TestSSHKnownHost(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.SyncSource, ntopts.SkipNonLocalGitProvider, ntopts.NamespaceRepo(backendNamespace, configsync.RepoSyncName))
+	knownHostValue := "dir"
+	var err error
+
+	// get known host key value
+	knownHostValue, err = nomostest.GetKnownHosts(nt)
+	if err != nil {
+		nt.T.Fatalf("generating kh: %s", knownHostValue)
+	}
+
+	// create secret with known_hosts key
+	rootSecret := &corev1.Secret{}
+	err = nt.Validate(nomostest.RootAuthSecretName, configsync.ControllerNamespace, rootSecret)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		core.RootReconcilerPrefix, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "false"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// apply known host key and validate
+	nt.MustMergePatch(rootSecret, secretDataPatch(controllers.KnownHostsKey, knownHostValue))
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		core.RootReconcilerPrefix, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "true"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// repo sync test
+	knownHostValue = "invalid value"
+	repoReconcilerName := core.NsReconcilerName(backendNamespace, configsync.RepoSyncName)
+	repoSecret := &corev1.Secret{}
+	err = nt.Validate(nomostest.NamespaceAuthSecretName, backendNamespace, repoSecret)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		repoReconcilerName, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "false"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.MustMergePatch(repoSecret, secretDataPatch(controllers.KnownHostsKey, knownHostValue))
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		repoReconcilerName, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "true"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	err = nt.Watcher.WatchObject(kinds.RepoSyncV1Beta1(), configsync.RepoSyncName, backendNamespace, []testpredicates.Predicate{
+		testpredicates.RepoSyncHasSourceError("2004", "No ED25519 host key is known"),
+	})
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Cleanup(func() {
```

Comment:
nit: optional: Cleanup is usually at the beginning of test

Reply:
not optional 😄 

the cleanup should be registered at the same time that the resources is created/updated

Reply:
moved cleanup to the beginning